2_run/src/run_glm3.R

```diff
@@ -0,0 +1,157 @@
+#'  meteo file name and hash for each model run
+#'  @param export_fl_template the template for constructing the filepath 
+#'  for the export feather file that will be saved in `extract_glm_ouput()`
+#'  @return a tibble which includes the run_date, lake_id, gcm, time_period, 
+#'  the name of the export feather file, its hash (NA if the model run failed), 
+#'  the duration of the model run, whether or not the model run succeeded, 
+#'  and the code returned by the call to GLM3r::run_glm(). 
+run_glm3_model <- function(sim_dir, nml_objs, meteo_xwalk, export_fl_template) {
+  # pull lake_id from meteo_xwalk
+  lake_id <- meteo_xwalk$site_id
+  time_period <- meteo_xwalk$time_period
+  gcm <- meteo_xwalk$gcm
+  
+  # prepare to write inputs and results locally for quick I/O
+  sim_lake_dir <- file.path(sim_dir, sprintf('%s_%s_%s', lake_id, gcm, time_period))
+  dir.create(sim_lake_dir, recursive=TRUE, showWarnings=FALSE)
+  # delete sim_lake_dir after model has run and we've extracted the results
+  on.exit(unlink(sim_lake_dir, recursive = TRUE))
+  
+  # Define time period begin and end dates
+  times <- strsplit(time_period,'_')[[1]]
+  time_period_begin <- sprintf('%s-01-01', times[1])
+  time_period_end <- sprintf('%s-12-31', times[2])
+  
+  # Read in meteo data, add burn in and burn out and save to sim_lake_dir
+  meteo_data <- munge_meteo_w_burn_in_out(meteo_xwalk, time_period_begin, time_period_end, 
+                                          burn_in = 300, burn_out = 190)
+  sim_meteo_filename <- 'meteo_fl.csv'
+  readr::write_csv(meteo_data, file.path(sim_lake_dir, sim_meteo_filename))
+  
+  # Define simulation start and stop dates using munged meteo data
+  sim_start <- format(meteo_data[1,]$time, "%Y-%m-%d")
+  sim_stop <- format(tail(meteo_data, 1L)$time, "%Y-%m-%d")
+  
+  # write nml file, specifying meteo file and start and stop dates:
+  nml_obj <- nml_objs[[lake_id]]
+  nml_obj <- set_nml(nml_obj, arg_list = list(nsave = 24, 
+                                              meteo_fl = sim_meteo_filename,
+                                              sim_name = sprintf('%s_%s_%s', lake_id, gcm, time_period),
+                                              start = sim_start,
+                                              stop = sim_stop))
+  nml_obj$light$Kw_file <- NULL # use static value instead of relying on a file
+  glmtools::write_nml(nml_obj, file.path(sim_lake_dir, 'glm3.nml'))
+  
+  # for each model run, try running the model up to 5 times
+  # if model run works, report how long that model took to run
+  # and extract output and save to feather export file
+  # return tibble with export file name and hash
+  # if model run fails after 5 attempts, return
+  # tibble indicating that it failed and giving the code
+  tryCatch(
+    {
+      retry::retry(
+        {glm_time <- system.time({glm_code <- GLM3r::run_glm(sim_lake_dir, verbose = FALSE)})[['elapsed']]},
```

Comment:
One additional thing I forgot to mention is that it is a really rare case where a model fails once but doesn't also fail the second, third, fourth, etc. time too when given identical inputs and config. I.e., GLM is a deterministic model. The rare case I've seen failures that weren't repeatable are when the compute itself had hiccups. 

Because of this, you may not need retries on the model. But having them won't hurt other than the 5 extra times a doomed model run would go through (which is probably actually quick).

Reply:
Ok, thanks for that info. I was wondering whether that was the case. In my initial testing, where I had a tryCatch that included a second try with verbose=TRUE, I was seeing that second try be triggered and be successful (after a failed first try) for a few model runs. Hence why I wanted to build in a few retry() calls. But that may have been when I still had a few fixes to make in the nml files, so might not have been a realistic scenario.